File: frontend/lib/api/customers.ts
import { apiClient } from './client'

export interface Customer {
  id: number
  business_id: number
  branch_id?: number
  name: string
  phone?: string
  telegram_user_id?: number
  email?: string
  address?: string
  notes?: string
  total_credit: number
  total_paid: number
  balance: number
  loyalty_points: number
  is_active: boolean
  is_settled: boolean
  created_at: string
  updated_at: string
  last_payment_at?: string
  last_sale_at?: string
}

export interface CustomerCreate {
  name: string
  phone?: string
  telegram_user_id?: number
  email?: string
  address?: string
  notes?: string
}

export interface CustomerUpdate {
  name?: string
  phone?: string
  telegram_user_id?: number
  email?: string
  address?: string
  notes?: string
}

export interface CreditEntry {
  id: number
  customer_id: number
  entry_type: 'credit' | 'payment'
  amount: number
  balance_after: number
  payment_method?: string
  reference?: string
  notes?: string
  created_at: string
}

export interface PaymentEntryCreate {
  customer_id: number
  amount: number
  payment_method: string
  reference?: string
  notes?: string
}

export interface CreditEntryCreate {
  customer_id: number
  amount: number
  sale_id?: number
  invoice_id?: number
  reference?: string
  notes?: string
}

export interface CustomerLedger {
  customer: Customer
  entries: CreditEntry[]
  current_balance: number
}

export interface AgingBucket {
  bucket: string
  count: number
  total_amount: number
  customers: Customer[]
}

export interface AgingReport {
  total_outstanding: number
  buckets: AgingBucket[]
}

export interface LoyaltyEntry {
  id: number
  customer_id: number
  entry_type: 'earned' | 'redeemed'
  points: number
  points_after: number
  redemption_type?: string
  redemption_value?: number
  notes?: string
  created_at: string
}

export const customersApi = {
  // List customers
  async getCustomers(search?: string): Promise<Customer[]> {
    const response = await apiClient.get<Customer[]>('/customers', {
      params: search ? { search } : {},
    })
    return response.data
  },

  // Get customer by ID
  async getCustomer(id: number): Promise<Customer> {
    const response = await apiClient.get<Customer>(`/customers/${id}`)
    return response.data
  },

  // Create customer
  async createCustomer(data: CustomerCreate): Promise<Customer> {
    const response = await apiClient.post<Customer>('/customers', data)
    return response.data
  },

  // Update customer
  async updateCustomer(id: number, data: CustomerUpdate): Promise<Customer> {
    const response = await apiClient.put<Customer>(`/customers/${id}`, data)
    return response.data
  },

  // Delete customer (soft delete)
  async deleteCustomer(id: number): Promise<void> {
    await apiClient.delete(`/customers/${id}`)
  },

  // Add credit entry
  async addCredit(data: CreditEntryCreate): Promise<CreditEntry> {
    const response = await apiClient.post<CreditEntry>('/customers/credit/add', data)
    return response.data
  },

  // Record payment
  async addPayment(data: PaymentEntryCreate): Promise<CreditEntry> {
    const response = await apiClient.post<CreditEntry>('/customers/payment/add', data)
    return response.data
  },

  // Get customer ledger
  async getLedger(customerId: number, limit?: number): Promise<CustomerLedger> {
    const response = await apiClient.get<CustomerLedger>(`/customers/${customerId}/ledger`, {
      params: limit ? { limit } : {},
    })
    return response.data
  },

  // Get aging report
  async getAgingReport(): Promise<AgingReport> {
    const response = await apiClient.get<AgingReport>('/credit/aging')
    return response.data
  },

  // Get customer loyalty history
  async getLoyaltyHistory(customerId: number, limit?: number): Promise<LoyaltyEntry[]> {
    const response = await apiClient.get<LoyaltyEntry[]>(`/customers/${customerId}/loyalty`, {
      params: limit ? { limit } : {},
    })
    return response.data
  },
}

File: backend/app/api/customers.py
"""
Customer API endpoints
"""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlmodel import Session
from typing import Optional, List
from app.api.dependencies import get_db, get_current_user
from app.models.user import User
from app.models.business import Business
from app.services.business import get_business_by_user_id
from app.services.customer_service import (
    create_customer,
    get_customer,
    get_customers_by_business,
    update_customer,
    delete_customer,
    get_customer_by_phone,
    get_customer_by_telegram_id
)
from app.services.credit_service import (
    add_credit_entry,
    add_payment_entry,
    get_customer_ledger
)
from app.services.loyalty_service import (
    add_loyalty_entry,
    get_customer_loyalty_history
)
from app.schemas.customer import (
    CustomerCreate,
    CustomerUpdate,
    CustomerResponse,
    CreditEntryCreate,
    PaymentEntryCreate,
    CreditEntryResponse,
    CustomerLedgerResponse,
    LoyaltyEntryCreate,
    LoyaltyEntryResponse
)

router = APIRouter(prefix="/customers", tags=["customers"])


@router.post("", response_model=CustomerResponse)
async def create_customer_endpoint(
    customer_data: CustomerCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Create a new customer"""
    business = get_business_by_user_id(db, current_user.id)
    if not business:
        raise HTTPException(status_code=404, detail="Business not found")
    
    # Check if customer with same phone already exists
    if customer_data.phone:
        existing = get_customer_by_phone(db, business.id, customer_data.phone)
        if existing:
            raise HTTPException(status_code=400, detail="Customer with this phone number already exists")
    
    customer = create_customer(db, business.id, customer_data, current_user.branch_id)
    return customer


@router.get("", response_model=List[CustomerResponse])
async def list_customers_endpoint(
    search: Optional[str] = Query(None),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """List all customers for the business"""
    business = get_business_by_user_id(db, current_user.id)
    if not business:
        raise HTTPException(status_code=404, detail="Business not found")
    
    customers = get_customers_by_business(db, business.id, current_user.branch_id, search)
    return customers


@router.get("/{customer_id}", response_model=CustomerResponse)
async def get_customer_endpoint(
    customer_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get customer details"""
    customer = get_customer(db, customer_id)
    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")
    
    business = get_business_by_user_id(db, current_user.id)
    if customer.business_id != business.id:
        raise HTTPException(status_code=403, detail="Access denied")
    
    return customer


@router.put("/{customer_id}", response_model=CustomerResponse)
async def update_customer_endpoint(
    customer_id: int,
    customer_data: CustomerUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Update customer information"""
    customer = get_customer(db, customer_id)
    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")
    
    business = get_business_by_user_id(db, current_user.id)
    if customer.business_id != business.id:
        raise HTTPException(status_code=403, detail="Access denied")
    
    updated = update_customer(db, customer_id, customer_data)
    return updated


@router.delete("/{customer_id}")
async def delete_customer_endpoint(
    customer_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Delete customer (soft delete)"""
    customer = get_customer(db, customer_id)
    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")
    
    business = get_business_by_user_id(db, current_user.id)
    if customer.business_id != business.id:
        raise HTTPException(status_code=403, detail="Access denied")
    
    delete_customer(db, customer_id)
    return {"success": True}


@router.post("/credit/add", response_model=CreditEntryResponse)
async def add_credit_endpoint(
    credit_data: CreditEntryCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Add a credit entry (sale on credit)"""
    business = get_business_by_user_id(db, current_user.id)
    if not business:
        raise HTTPException(status_code=404, detail="Business not found")
    
    try:
        entry = add_credit_entry(db, business.id, current_user.id, credit_data)
        return entry
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/payment/add", response_model=CreditEntryResponse)
async def add_payment_endpoint(
    payment_data: PaymentEntryCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Record a payment against customer credit"""
    business = get_business_by_user_id(db, current_user.id)
    if not business:
        raise HTTPException(status_code=404, detail="Business not found")
    
    try:
        entry = add_payment_entry(db, business.id, current_user.id, payment_data)
        return entry
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/{customer_id}/ledger", response_model=CustomerLedgerResponse)
async def get_customer_ledger_endpoint(
    customer_id: int,
    limit: int = Query(100, le=500),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get customer credit ledger"""
    customer = get_customer(db, customer_id)
    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")
    
    business = get_business_by_user_id(db, current_user.id)
    if customer.business_id != business.id:
        raise HTTPException(status_code=403, detail="Access denied")
    
    entries = get_customer_ledger(db, customer_id, limit)
    
    return CustomerLedgerResponse(
        customer=customer,
        entries=entries,
        current_balance=customer.balance
    )


@router.post("/loyalty/add", response_model=LoyaltyEntryResponse)
async def add_loyalty_entry_endpoint(
    loyalty_data: LoyaltyEntryCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Add a loyalty entry (earned or redeemed)"""
    business = get_business_by_user_id(db, current_user.id)
    if not business:
        raise HTTPException(status_code=404, detail="Business not found")
    
    try:
        entry = add_loyalty_entry(db, business.id, current_user.id, loyalty_data)
        return entry
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/{customer_id}/loyalty", response_model=List[LoyaltyEntryResponse])
async def get_customer_loyalty_endpoint(
    customer_id: int,
    limit: int = Query(50, le=200),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get customer loyalty points history"""
    customer = get_customer(db, customer_id)
    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")
    
    business = get_business_by_user_id(db, current_user.id)
    if customer.business_id != business.id:
        raise HTTPException(status_code=403, detail="Access denied")
    
    entries = get_customer_loyalty_history(db, customer_id, limit)
    return entries

